Initial WASI/componentize-py support

mschuwalow opened the pull request:
<!---
Thanks for your contribution! ♥️

If this is your first PR to urllib3 please review the Contributing Guide:
https://urllib3.readthedocs.io/en/latest/contributing.html

Adhering to the Contributing Guide means we can review, merge, and release your change faster! :)
--->

resolves #3584 

This pr adds initial support for using urllib3 in a wasi / componentize-py environment.

This is achieved using a similar setup as the emscripten support, by monkeypatching the connection classes when we are in a wasi environment.

The support is not perfect, some things that are normally done by the library such as transfer encoding are done by wasi-http / the host now, so the options will have no effect. It should still lower the barrier of entry for users to use python with wasi significantly, so I think there is value here.

mschuwalow commented on src/urllib3/contrib/wasi/wasi.py:
```diff
@@ -0,0 +1,159 @@
+    importlib.import_module(".imports.streams", world_name)
+    importlib.import_module(".types", world_name)
+
+def send_request(request: WasiRequest, world_name: str) -> WasiResponse:
```
This being untyped is a bit annoying, but I don't think there is a good workaround currently.

Hopefully componentize-py eventually allows generating the bindings in a static location so we can have type hints here

mschuwalow commented on src/urllib3/response.py:
```diff
@@ -9,7 +9,10 @@
-import zlib
+try:
+    import zlib
+except ImportError:
```
zlib is not currently available in wasi, so I guarded the import for now.
For all other systems there should not be a difference

mschuwalow committed:
initial wasi/componentize-py support

mschuwalow commented:
I'll update this to the latest componentize-py version once https://github.com/bytecodealliance/componentize-py/pull/150 is merged.

This will get rid of the `enable_wasi_backend` function / dynamic imports and will allow to have types in the wasi interaction code.

mschuwalow commented on src/wit_world/__init__.py:
```diff
@@ -0,0 +1,13 @@
+from typing import TypeVar, Generic, Union, Optional, Protocol, Tuple, List, Any, Self
```
This is generated by the wasi-tools/regenerate_bindings.sh script and excluded from packaging. At runtime this will be provided by componentize-py on wasi

mschuwalow commented on pyproject.toml:
```diff
@@ -129,6 +133,9 @@ include = [
   "/README.md",
   "/LICENSE.txt",
 ]
+exclude = [
```
So the generated code is not packaged

mschuwalow commented on wasi-tools/regenerate_bindings.sh:
```diff
@@ -0,0 +1,4 @@
+#! /bin/sh
+
+rm -rf src/wit_world
```
This generates the code in src/wit_world and needs to be rerun when the code in the wasi-tools/wit directory changes

mschuwalow commented on wasi-tools/wit/deps.toml:
```diff
@@ -0,0 +1,4 @@
+io = "https://github.com/WebAssembly/wasi-io/archive/v0.2.0.tar.gz"
```
The wit files are fetched using the [wit-deps](https://github.com/bytecodealliance/wit-deps) tool.
Everything in wasi-tools/wit/deps/* is generated by it.

rajsite commented:
@mschuwalow I'm really interested in this support, are there any substantial blockers left?

mschuwalow commented:
> @mschuwalow I'm really interested in this support, are there any substantial blockers left?

As far as I'm aware this should be working with the latest componentize-py without any known issues (though it was only use for a few smaller demos so far). Will rebase this in the next few days

crepererum committed:
fix: return cached result from `WasiHttpResponseWrapper.stream` (#2)

`requests` will call `stream` and expects data to be present there, so
we need to return the cached result. Otherwise `requests` will only try
to poll the exhausted stream and will not get any data, resulting in an
empty body.

crepererum committed:
fix: correctly stream large responses under WASI (#3)

Given the following code and a rather large response (>=10k), the code
prior to this change would only buffer up the first chunk/part of the
response. This happens due to a disagreement about `preload_content`
between `urllib3.contrib.wasi.wasi.convert_body` and
`urllib3.contrib.wasi.response.WasiHttpResponseWrapper`.

```python
import urllib3

def perform_request(url: str) -> str:
    resp = urllib3.request(
        "GET",
        url,
        preload_content=False,
        decode_content=False,
    )

    data = b""
    for part in resp.stream():
        data += part
    return data.decode("utf-8")
```

This may affect all users but especially `requests`, which uses
`preload_content=False` and `decode_content=False`.

crepererum committed:
fix: `ACCEPT_ENCODING` & `make_headers` if `zlib` is missing (#4)

CPython can be compiled without zlib, e.g. for WASI targets. This was
already considered for the decompression routines in `urllib3.response`,
but the `Accept-Encoding` header in `urllib3.util.request` always added
`gzip` & `deflate` to the set of accepted encodings. Esp. in combination
with `requests`, that had the unexpected outcome that
`requests.get(...).text` returned the compressed server response instead
of usable decompressed content.

mschuwalow committed:
Merge remote-tracking branch 'origin/main' into wasi

mschuwalow committed:
pin action

mschuwalow committed:
format code

mschuwalow committed:
Merge branch 'main' into wasi

mschuwalow committed:
Merge remote-tracking branch 'origin/main' into wasi

mschuwalow committed:
Merge remote-tracking branch 'origin/main' into wasi

mschuwalow committed:
update lockfile
